account/identity_management.go

```diff
@@ -0,0 +1,180 @@
+package account
+
+import (
+	"context"
+	uuid "github.com/satori/go.uuid"
+)
+
+const (
+	IdentityResourceTypeOrganization = "identity/organization"
+	IdentityResourceTypeTeam         = "identity/team"
+	IdentityResourceTypeGroup        = "identity/group"
+	IdentityResourceTypeUser         = "identity/user"
+)
+
+// This struct is used to return the Organizations for which an Identity is associated
+type IdentityOrganization struct {
+	OrganizationID uuid.UUID
+	Name           string
+	Member         bool
+	Roles          []string
+}
+
+// Returns an array of all organizations in which the specified user is a member or is assigned a role
+func (m *GormIdentityRepository) ListOrganizations(ctx context.Context, identityID uuid.UUID) ([]IdentityOrganization, error) {
```

Comment:
I think there's a test missing for this function.

Reply:
It's tested as part of `controller/organization_blackbox_test.go` - since it's non-trivial to actually create an organization without using the controller anyway, it doesn't make sense to try and test this function separately.

Reply:
The tests `controller/organization_blackbox_test.go` is to test the controller irrespective of whether we are using the actual or mocked 'dao' /repository layer.

However the tests here at the repository / dao layer is to ensure that this `ListOrganizations` Golang API method works as per the contract - example, the tests in https://github.com/fabric8-services/fabric8-auth/blob/master/account/identity_blackbox_test.go

Reply:
So how would you suggest we create the organization?  Do we use the controller's Create() method? (I didn't know if we have access to the controllers from these tests?)

Reply:
You could use repository methods or SQLs to do so - but usually we use the repository methods since they wrap the SQLs.
The way I did here https://github.com/fabric8-services/fabric8-auth/pull/322/files#diff-4bd1c185cd704de02aaa93a028975a89R50

Reply:
Yes but as I said it's non-trivial to actually create an organization - you need to create a resource object, an identity object, assign a role to the user creating it, and so on.. just look at all the code in the organization controller's Create() method.  Are you suggesting I replicate all of this complex functionality in the repository test to create an organization simply so that I have data to test the ListOrganizations() method?  In this case, I would have thought that the integration tests would exercise this code sufficiently, especially since there are numerous other tests still to come once we implement organization/team/group memberships.

Reply:
> Are you suggesting I replicate all of this complex functionality in the repository test to create an organization simply so that I have data to test the ListOrganizations() method? 

Yes, because the controller tests are for validating the controller contract. The repository tests are for validating the repository contract in a specific package. The tests are critical here because the code in this package/repository is going to consumed by others packages.

Reply:
I don't like the idea of duplicating so much code, how about I look at some way of moving the create organization logic to the repository codebase instead of having it in the controller?  The only question I have is whether it is ok to have a repository method access other repositories/tables.  To me this seems allowable, as the ListOrganizations() method is already querying multiple tables anyway.  What do you think?

Reply:
>  how about I look at some way of moving the create organization logic to the repository codebase instead of having it in the controller? 

That's a great idea irrespective of our conversation around tests. We should abstract out `CreateOrganization(..)` to a repository method !

In fact, ( contradictory to a few comments back ) I'm even wondering if we should go with an `OrganizationRepository` which has the two methods `Create(..)` , `List(..)` ? Looks a lot cleaner to me , that way :)

I am thinking along the lines of having `account/organization.go` .. which would have the  above `OrganizationRepository` definition & methods.

Reply:
Or you could define an `OrganizationService` in `account/organization.go` with `List` and `Create` options.

Reply:
I like the idea of creating a new repository, however how about we call it IdentityManagementRepository and have it include the team/group and membership operations provided there also, as these are all very closely related?

Reply:
Making it a service object is appealing also, as it could simply interact with the existing repositories / directly with the db via queries.

Reply:
The issue I see with having different files for each thing, is that there are an awful lot of cross-cutting concerns here.  Assigning roles, managing memberships, etc all apply to organizations, teams and groups, not to mention querying permissions applies to all of them also.  If we don't want to group them together just because we don't like "IdentityManagement" as a name, is there an alternative name that we could use instead?

Reply:
Ok, I've thought about this some more.  Perhaps we could have a separate service file for each type of user collection, i.e. OrganizationService, TeamService, GroupService, and then in addition to this have services for each of the cross-cutting concerns, e.g. RoleService, MembershipService, InvitationService, etc.  I think the number of new files we would be creating is reasonable this way.  Also, do you think we would access these services in the same way that we access the repositories? i.e. via the application object?  We could easily provide application.GetOrganizationService(), application.GetTeamService() and so on, which would set up the service objects with the database connection.

One last thought - placing these in the account package doesn't seem intuitive to me, as these features aren't really account-related.  It is a pity that the "account" package isn't called "identity" instead, then it would make more sense to put them there.  Do you have any suggestions as to the best place to locate these services?

Reply:
Fair concerns @sbryzak . 

I was thinking along the lines of how we would design the structs/interfaces irrespective of splitting files.

```
type OrganizationService interface {
         List(...)
         Create(...) // calls of a bunch of repository methods from roles/resources/identities         
         AssignRole(..) // calls `RoleRepository` methods
         GetPermissions(..)
}
```

```
type TeamService interface {
         List(...)  
         Create(...) // calls of a bunch of repository methods from roles/resources/identities
         AssignRole(..) // calls `RoleRepository` methods
         GetPermissions(..)
}
```


```
type GroupService interface {
         List(...)  
         Create(...) // calls of a bunch of repository methods from roles/resources/identities
         AssignRole(..) // calls `RoleRepository` methods
         GetPermissions(..)
}
```

I haven't thought it out entirely but the above design is something we need to come up with - we could put them in the same/different file/package based on the dependencies :)

Reply:
Looks like we commented at the same time :) I agree with your *entire* comment and the details could best be picked/improvised while implementing.

> One last thought - placing these in the account package doesn't seem intuitive to me, as these features aren't really account-related. It is a pity that the "account" package isn't called "identity" instead, then it would make more sense to put them there. Do you have any suggestions as to the best place to locate these services?

Was wondering so too.. let me think.

Reply:
Maybe under sub packages in `authorization` ? :)

Reply:
Ah, how about the `authorization` package itself?

Reply:
Yup, should be fine :)